Binary incompatibility with Mockito 2.4.0

Mockito 2.4.0 has removed the class `org.mockito.internal.runners.util.FrameworkUsageValidator`. This means Jukito can only be used with Mockito 2.3.11 or earlier for now.

Reproducible with any project that combines Jukito and Mockito 2.4.x, for example, clone https://github.com/aalmiray/javatrove/tree/master/github-api-01 and change `mockitoVersion` in `gradle.properties`

Full stacktrace follows

    java.lang.NoClassDefFoundError: org/mockito/internal/runners/util/FrameworkUsageValidator
        at java.lang.Class.getDeclaredConstructors0(Native Method)
        at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
        at java.lang.Class.getConstructor0(Class.java:3075)
        at java.lang.Class.getConstructor(Class.java:1825)
        at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
        at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
        at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
        at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
        at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
        at org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:33)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:86)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:57)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
        at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
        at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:377)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
    Caused by: java.lang.ClassNotFoundException: org.mockito.internal.runners.util.FrameworkUsageValidator
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 36 more

--- a/jukito/src/main/java/org/jukito/JukitoRunner.java
+++ b/jukito/src/main/java/org/jukito/JukitoRunner.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright 2013 ArcBees Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not
@@ -34,7 +34,6 @@ import org.junit.runners.BlockJUnit4ClassRunner;
 import org.junit.runners.model.FrameworkMethod;
 import org.junit.runners.model.InitializationError;
 import org.junit.runners.model.Statement;
-import org.mockito.internal.runners.util.FrameworkUsageValidator;
 
 import java.lang.annotation.Annotation;
 import java.lang.reflect.InvocationTargetException;
@@ -45,24 +44,19 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
-/**
- * TODO: Rework this documentation
- * <p/>
+/*
  * This class implements the mockito runner but allows Guice dependency
  * injection. To setup the guice environment, the test class can have an inner
  * static class deriving from {@link TestModule}. This last class will let you bind
  * {@link TestSingleton} and {@link TestEagerSingleton} and the runner will make sure these
  * singletons are reset at every invocation of a test case.
- * <p/>
- * This code not very clean as it is cut & paste from
- * {@link org.mockito.internal.runners.JUnit45AndHigherRunnerImpl}, but it's
- * unclear how we could make otherwise.
- * <p/>
+ *
+ *
  * Most of the code here is inspired from: <a href=
  * "http://cowwoc.blogspot.com/2008/10/integrating-google-guice-into-junit4.html"
  * > http://cowwoc.blogspot.com/2008/10/integrating-google-guice-into-junit4.
  * html</a>
- * <p/>
+ *
  * Depends on Mockito.
  */
 public class JukitoRunner extends BlockJUnit4ClassRunner {
@@ -156,7 +150,7 @@ public class JukitoRunner extends BlockJUnit4ClassRunner {
      */
     private Set<Class<? extends Module>> getUseModuleClasses(Class<?> testClass) {
         Class<?> currentClass = testClass;
-        Set<Class<? extends Module>> modules = new HashSet<Class<? extends Module>>();
+        Set<Class<? extends Module>> modules = new HashSet<>();
         while (currentClass != null) {
             UseModules useModules = currentClass.getAnnotation(UseModules.class);
             if (useModules != null) {
@@ -229,7 +223,7 @@ public class JukitoRunner extends BlockJUnit4ClassRunner {
     @Override
     public void run(RunNotifier notifier) {
         // add listener that validates framework usage at the end of each test
-        notifier.addListener(new FrameworkUsageValidator(notifier));
+        notifier.addListener(new MockitoUsageValidator(notifier));
         super.run(notifier);
     }
 
@@ -319,11 +313,11 @@ public class JukitoRunner extends BlockJUnit4ClassRunner {
      * Computes a list of all bindings which match a {@link All} annotation.
      *
      * @param allAnnotation the annotation to match
-     * @param typeLiteral the type of the bindings.
+     * @param typeLiteral   the type of the bindings.
      * @return the computed list.
      */
     private List<Binding<?>> getBindingsForParameterWithAllAnnotation(All allAnnotation, TypeLiteral<?> typeLiteral) {
-        List<Binding<?>> result = new ArrayList<Binding<?>>();
+        List<Binding<?>> result = new ArrayList<>();
         String bindingName = allAnnotation.value();
         if (All.DEFAULT.equals(bindingName)) {
             // If the annotation is with the default name bind all bindings
@@ -357,7 +351,7 @@ public class JukitoRunner extends BlockJUnit4ClassRunner {
             Method javaMethod, List<FrameworkMethod> result) {
 
         if (index >= bindingsToUseForParameters.size()) {
-            List<Binding<?>> assignation = new ArrayList<Binding<?>>(currentAssignation.size());
+            List<Binding<?>> assignation = new ArrayList<>(currentAssignation.size());
             assignation.addAll(currentAssignation);
             result.add(new InjectedFrameworkMethod(javaMethod, assignation));
             return;
@@ -422,6 +416,7 @@ public class JukitoRunner extends BlockJUnit4ClassRunner {
      * Adds to {@code errors} for each method annotated with {@code @Test}that
      * is not a public, void instance method with no arguments.
      */
+    @Override
     protected void validateTestMethods(List<Throwable> errors) {
         validatePublicVoidMethods(Test.class, false, errors);
     }
@@ -435,8 +430,8 @@ public class JukitoRunner extends BlockJUnit4ClassRunner {
      * <li>is static (given {@code isStatic is false}), or
      * <li>is not static (given {@code isStatic is true}).
      */
-    protected void validatePublicVoidMethods(Class<? extends Annotation> annotation,
-                                             boolean isStatic, List<Throwable> errors) {
+    private void validatePublicVoidMethods(Class<? extends Annotation> annotation,
+                                           boolean isStatic, List<Throwable> errors) {
         List<FrameworkMethod> methods = getTestClass().getAnnotatedMethods(annotation);
 
         for (FrameworkMethod eachTestMethod : methods) {

--- a/jukito/src/main/java/org/jukito/MockitoUsageValidator.java
+++ b/jukito/src/main/java/org/jukito/MockitoUsageValidator.java
@@ -1,0 +1,41 @@
+/*
+ * Copyright 2017 ArcBees Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.jukito;
+
+import org.junit.runner.notification.Failure;
+import org.junit.runner.notification.RunListener;
+import org.junit.runner.notification.RunNotifier;
+import org.mockito.Mockito;
+
+public class MockitoUsageValidator extends RunListener {
+    private final RunNotifier notifier;
+
+    public MockitoUsageValidator(RunNotifier notifier) {
+        this.notifier = notifier;
+    }
+
+    @Override
+    public void testFinished(org.junit.runner.Description description) throws Exception {
+        super.testFinished(description);
+
+        try {
+            Mockito.validateMockitoUsage();
+        } catch (Throwable t) {
+            notifier.fireTestFailure(new Failure(description, t));
+        }
+    }
+}

--- a/pom.xml
+++ b/pom.xml
@@ -130,7 +130,7 @@
 
         <guice.version>4.0</guice.version>
         <junit.version>4.12</junit.version>
-        <mockito.version>1.10.8</mockito.version>
+        <mockito.version>2.5.7</mockito.version>
 
         <maven-checkstyle-plugin.version>2.10</maven-checkstyle-plugin.version>
         <maven-compiler-plugin.version>3.0</maven-compiler-plugin.version>

